latch_cli/services/cp/download.py

```diff
@@ -0,0 +1,176 @@
+                show_total_progress=show_total_progress,
+                verbose=config.verbose,
+            )
+            progress_bars.set_total(num_files, "Copying Files")
+
+            start = time.perf_counter()
+            with ProcessPoolExecutor() as executor:
+                total_bytes = sum(
+                    executor.map(
+                        download_file,
+                        urls,
+                        dests,
+                        repeat(config.chunk_size),
+                        repeat(progress_bars),
+                    )
+                )
+            end = time.perf_counter()
+            progress_bars.close()
+    else:
+        file_data: GetSignedUrlData = json_data["data"]
+
+        num_files = 1
+
+        if dest.exists() and dest.is_dir():
+            dest = dest / node_data.name
+
+        if config.progress == Progress.none:
+            num_bars = 0
+        else:
+            num_bars = 1
+
+        with ProgressBarManager() as manager:
+            progress_bars: ProgressBars = manager.ProgressBars(
+                num_bars,
+                show_total_progress=False,
+                verbose=config.verbose,
+            )
+
+            start = time.perf_counter()
+            total_bytes = download_file(
+                file_data["url"], dest, config.chunk_size, progress_bars
+            )
+            end = time.perf_counter()
+            progress_bars.close()
+
+    total_time = end - start
+
+    click.secho(
+        (
+            "Downloaded"
+            f" {num_files} {pluralize('file', 'files', num_files)} ({with_si_suffix(total_bytes)})"
+            f" in {human_readable_time(total_time)}."
+        ),
+        fg="green",
+    )
+
+
+# dest will always be a path which includes the copied file as its leaf
+# e.g. download_file("a/b.txt", Path("c/d.txt")) will copy the content of 'b.txt' into 'd.txt'
+def download_file(
+    signed_url: str,
+    dest: Path,
+    chunk_size: int,
+    progress_bars: ProgressBars,
+) -> int:
+    dest.parent.mkdir(parents=True, exist_ok=True)
+    with open(dest, "wb") as f:
+        res = tinyrequests.get(signed_url, stream=True)
+
+        total_bytes = res.headers.get("Content-Length")
+        if total_bytes is not None:
+            pbar_index = progress_bars.get_free_task_bar_index()
+            progress_bars.set(index=pbar_index, total=int(total_bytes), desc=dest.name)
+
+            for data in res.iter_content(chunk_size):
+                f.write(data)
+                progress_bars.update(pbar_index, len(data))
+
+            progress_bars.return_task_bar(pbar_index)
+            progress_bars.update_total_progress(1)
+            progress_bars.write(f"Copied {dest.name}")
+
+            return int(total_bytes)
+        else:
+            total_bytes = 0
+            for data in res.iter_content(chunk_size):
+                f.write(data)
+                total_bytes += len(data)
+
+            return total_bytes
```

Comment:
Is it possible to do multipart concurrent downloads using range queries in the pre-signed URLs? Not saying for right now.

Reply:
added a benchmark todo